pycqed/instrument_drivers/meta_instrument/heterodyne.py

```diff
@@ -23,75 +23,295 @@ class HeterodyneInstrument(Instrument):
-                           vals=vals.Numbers(9e3, 40e9))
-        self.add_parameter('f_RO_mod', parameter_class=ManualParameter,
-                           vals=vals.Numbers(-600e6, 600e6),
-                           label='Intermodulation frequency',
-                           units='Hz', initial_value=10e6)
+
+        self.common_init(name, LO, AWG, acquisition_instr,
+                         single_sideband_demod, **kw)
+
         self.add_parameter('RF_power', label='RF power',
-                           units='dBm',
+                           units='dBm', vals=vals.Numbers(),
                            set_cmd=self.do_set_RF_power,
                            get_cmd=self.do_get_RF_power)
-        self.add_parameter('single_sideband_demod',
-                           label='Single sideband demodulation',
-                           parameter_class=ManualParameter)
-        self.add_parameter('acquisition_instr',
-                           set_cmd=self._do_set_acquisition_instr,
-                           get_cmd=self._do_get_acquisition_instr,
-                           vals=vals.Anything())
         self.add_parameter('acquisition_instr_controller',
-                           set_cmd=self._do_set_acquisition_instr_controller,
-                           get_cmd=self._do_get_acquisition_instr_controller,
+                           set_cmd=self.do_set_acquisition_instr_controller,
+                           get_cmd=self.do_get_acquisition_instr_controller,
                            vals=vals.Anything())
-        self.add_parameter('nr_averages',
+        self.acquisition_instr_controller(acquisition_instr_controller)
+        self._RF_power = None
+
+    def common_init(self, name, LO, AWG, acquisition_instr='CBox',
+                    single_sideband_demod=False, **kw):
+        logging.info(__name__ + ' : Initializing instrument')
+        Instrument.__init__(self, name, **kw)
+
+        self.LO = LO
+        self.AWG = AWG
+        self.add_parameter('frequency', label='Heterodyne frequency',
+                           units='Hz', vals=vals.Numbers(9e3, 40e9),
+                           get_cmd=self.do_get_frequency,
+                           set_cmd=self.do_set_frequency)
+        self.add_parameter('f_RO_mod', label='Intermodulation frequency',
+                           units='Hz', vals=vals.Numbers(-600e6, 600e6),
+                           set_cmd=self.do_set_f_RO_mod,
+                           get_cmd=self.do_get_f_RO_mod)
+        self.add_parameter('single_sideband_demod', vals=vals.Bool(),
+                           label='Single sideband demodulation',
+                           parameter_class=ManualParameter,
+                           initial_value=single_sideband_demod)
+        self.add_parameter('acquisition_instr', vals=vals.Strings(),
+                           label='Acquisition instrument',
+                           set_cmd=self.do_set_acquisition_instr,
```

Comment:
this should be a private command (starting with _). I believe it was like this in the last version.